ambry-commons/src/main/java/com.github.ambry.commons/ByteBufferReadableStreamChannel.java

```diff
@@ -65,6 +74,30 @@ public long getSize() {
   }
 
   @Override
+  public void setDigestAlgorithm(String digestAlgorithm)
+      throws NoSuchAlgorithmException {
+    if (digest != null && digest.getAlgorithm().equals(digestAlgorithm)) {
+      return;
+    }
+    digest = MessageDigest.getInstance(digestAlgorithm);
+    digestBytes = null;
+  }
+
+  @Override
+  public byte[] getDigest() {
+    if (digest == null) {
+      return null;
+    } else if (digestBytes == null) {
+      int savedPosition = buffer.position();
```

Comment:
Looks like these can be made into a utility method (calculating digest for a buffer given the digestAlgo) so that all your implementation can call into it wherever required instead of copy pasting the same 4 to 5 lines of code